libraries/AP_Compass/AP_Compass_RM3100.cpp

```diff
@@ -0,0 +1,208 @@
+#define CCP1    0x00
+#define GAIN_CC50 20.0f   // LSB/uT
+#define GAIN_CC100 38.0f
+#define GAIN_CC200 75.0f
+#define UTESLA_TO_MGAUSS   10.0f // uT to mGauss conversion
+
+#define TMRC    0x94    // Update rate 150Hz
+#define CMM     0x71    // read 3 axes and set data ready if 3 axes are ready
+
+extern const AP_HAL::HAL &hal;
+
+AP_Compass_Backend *AP_Compass_RM3100::probe(AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                              bool force_external,
+                                              enum Rotation rotation)
+{
+    if (!dev) {
+        return nullptr;
+    }
+    AP_Compass_RM3100 *sensor = new AP_Compass_RM3100(std::move(dev), force_external, rotation);
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_RM3100::AP_Compass_RM3100(AP_HAL::OwnPtr<AP_HAL::Device> _dev,
+                                       bool _force_external,
+                                       enum Rotation _rotation)
+    : dev(std::move(_dev))
+    , force_external(_force_external)
+    , rotation(_rotation)
+{
+}
+
+bool AP_Compass_RM3100::init()
+{
+    if (!dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        return false;
+    }
+
+    // high retries for init
+    dev->set_retries(10);
+
+    // maybe add a test for revision id value here and goto fail if not correct
+    // it is not clear if revision id is the same for all compasses, in the same manner as a whoami
+
+    dev->setup_checked_registers(8);
+
+    dev->write_register(RM3100_TMRC_REG, TMRC, true); // cycle count z
+    dev->write_register(RM3100_CMM_REG, CMM, false); // CMM configuration
+    dev->write_register(RM3100_CCX1_REG, CCP1, true); // cycle count x
+    dev->write_register(RM3100_CCX0_REG, CCP0, true); // cycle count x
+    dev->write_register(RM3100_CCY1_REG, CCP1, true); // cycle count y
+    dev->write_register(RM3100_CCY0_REG, CCP0, true); // cycle count y
+    dev->write_register(RM3100_CCZ1_REG, CCP1, true); // cycle count z
+    dev->write_register(RM3100_CCZ0_REG, CCP0, true); // cycle count z
+
+    _scaler = (1 / GAIN_CC200) * UTESLA_TO_MGAUSS; // has to be changed if using a different cycle count
+
+    // lower retries for run
+    dev->set_retries(3);
+    
+    dev->get_semaphore()->give();
+
+    /* register the compass instance in the frontend */
+    compass_instance = register_compass();
+
+    printf("Found a RM3100 at address 0x%x as compass %u\n", dev->get_bus_address(), compass_instance);
+    
+    set_rotation(compass_instance, rotation);
+
+    if (force_external) {
+        set_external(compass_instance, true);
+    }
+    
+    dev->set_device_type(DEVTYPE_RM3100);
+    set_dev_id(compass_instance, dev->get_bus_id());
+
+    // call timer() at 80Hz
+    dev->register_periodic_callback(1000000U/80U,
+                                    FUNCTOR_BIND_MEMBER(&AP_Compass_RM3100::timer, void));
+
+    return true;
+
+//fail:
```

Comment:
better to remove this unused code

Reply:
Ok. I will uncomment it when I add the whoami substitute.